server/routerlicious/packages/services-shared/src/redisSocketIoAdapter.ts

```diff
@@ -0,0 +1,506 @@
+    subConnection: ISocketIoRedisSubscriptionConnection;
+
+    // when set, enables per room health checks. messages are periodically published
+    healthChecks?: {
+        // how often to health check each room in milliseconds
+        interval: number,
+
+        // how long to wait for a health check to complete before failing it in milliseconds
+        timeout: number;
+
+        // determines if the adapter should resubscribe to the room if a health check fails
+        resubscribeOnFailure: boolean;
+
+        // called when a health check succeeds or fails. useful for telemetry purposes
+        onHealthCheck?(callerId: string, startTime: number, error?: any): void;
+    };
+
+    // called when recieving a message. useful for telemetry purposes
+    onReceive?(channel: string, startTime: number, packet: any, error?: any): void;
+}
+
+/**
+ * Custom version of the socket.io-redis adapter
+ * Differences between this and socket.io-redis:
+ * - Creates per room subscriptions which significantly reduces Redis server load for
+ * Fluid scenarios when running a large amount of Fluid frontend servers.
+ * - Contains a health checker that verifies each room is works *
+ * - Disables rooms for the default "/" namespace to reduce memory usage
+ * (https://github.com/socketio/socket.io/issues/3089)
+ * - Callbacks for telemetry logging
+ * The Redis pubsub channels are compatible with socket.io-redis
+ * References:
+ * - https://github.com/socketio/socket.io-redis
+ * - https://github.com/socketio/socket.io-emitter
+ * - https://github.com/socketio/socket.io-adapter
+ */
+export class RedisSocketIoAdapter extends EventEmitter implements socketio.Adapter {
+    private static options: ISocketIoRedisOptions;
+
+    // required for socketio.Adapter typing - however these are not used within socketio
+    public rooms: any = undefined;
+    public sids: any = undefined;
+
+    /**
+     * Map of room id to socket ids
+     * Shows what sockets are in a given room
+     */
+    public readonly roomToSocketIds: Map<string, Set<string>> = new Map();
+
+    /**
+     * Map of socket id to room ids
+     * Shows what rooms the given socket is id
+     */
+    private readonly socketIdToRooms: Map<string, Set<string>> = new Map();
+
+    private _uniqueRoomCount = 0;
+
+    private readonly uid: string;
+    private readonly channel: string;
+    private readonly encoder: any;
+
+    private readonly pendingHealthChecks: Map<string, () => void> = new Map();
+    private readonly roomHealthCheckTimeoutIds: Map<string, NodeJS.Timeout> = new Map();
+
+    /**
+     * Set the Redis connections to use
+     */
+    public static async setup(options: ISocketIoRedisOptions) {
+        this.options = options;
+    }
+
+    constructor(public readonly nsp: socketio.Namespace) {
+        super();
+
+        this.encoder = (nsp.server as any).encoder;
+
+        // todo: better id here?
+        this.uid = uuid.v4().substring(0, 6);
```

Comment:
Is taking 6 chars just to save the memory? Or its a requirement?

Reply:
A UID is required to make it so the server ignores messages it published itself. We could change it to any length, but the socket.io-redis adapter used a length of 6 so I stuck with that